coalib/io/FileCache.py

```diff
@@ -0,0 +1,69 @@
+import os
+
+__all__ = ['get_content', 'clear_cache']
+
+
+# The cache. Maps filenames to either a thunk which will provide source code,
+# or a tuple (size, mtime, contents, fullname) once loaded.
+cache = {}
+
+
+def clear_cache():
+    """
+    Clear the cache entirely.
+    """
+    global cache
+    cache = {}
+
+
+def get_content(filename):
+    """
+    Get the raw contents of a file from the cache.
+    Update the cache if it doesn't contain an entry for this file already.
+    """
+    if filename in cache:
+        return cache[filename][2]
+    return update_cache(filename)
+
+
+def update_cache(filename):
+    """
+    Update a cache entry and return its list of lines.
+    If something's wrong, print a message, discard the cache entry,
+    and return an empty list.
+    """
+    stat = os.stat(filename)
+
+    with open(filename, 'rb') as fp:
+        lines = fp.read()
+    size, mtime = stat.st_size, stat.st_mtime
+    cache[filename] = size, mtime, lines, filename
+    return lines
+
+
+class Memoize:
+    """
+    ``Memoize`` serves as a decorator for the properties
+    of ``FileFactory`` and caches the results of the property
+    for a particular ``FileFactory`` object.
+    """
+
+    def __init__(self, method):
+        self._method = method
+
+    def __call__(self, instance):
+        self._instance = instance
+        return self.__get__(self._instance, self._instance.__class__)
+
+    def __get__(self, instance, owner):
```

Comment:
actually, do you need to override `__get__`? Can you explain me how it works exactly? I believe `__get__` will be invoked when you do `self.memoized_func.attribute_you_want_to_access` (and that will only happen if it's not a property, as a property will always *call* the Memoize class).

Reply:
The results of the properties are stored in `self._cache` which is tied to the factory object. So I override `__get__` to access that attribute to store the results in the cache. And then `self.__get__` is returned on every `__call__`

Reply:
But isn't it possible to call `__get__` differently? Like just `get_cache`? Why do we have to override the magic method `__get__`?